ignore.c

```diff
@@ -0,0 +1,117 @@
+/* Copyright 2016-present Facebook, Inc.
+ * Licensed under the Apache License, Version 2.0 */
+
+#include "watchman.h"
+
+// The path and everything below it is ignored.
+#define kFullIgnore  (void*)0x1
+// The grand-children of the path are ignored, but not the path
+// or its direct children.
+#define kVCSIgnore   (void*)0x2
+
+bool w_ignore_init(struct watchman_ignore *ignore) {
+  ignore->ignore_vcs = w_ht_new(2, &w_ht_string_funcs);
+  if (!ignore->ignore_vcs) {
+    return false;
+  }
+  ignore->ignore_dirs = w_ht_new(2, &w_ht_string_funcs);
+  if (!ignore->ignore_dirs) {
+    return false;
+  }
+  if (art_tree_init(&ignore->tree) != 0) {
+    return false;
+  }
+  return true;
+}
+
+void w_ignore_add(struct watchman_ignore *ignore, const char *path,
+                  uint32_t pathlen, bool is_vcs_ignore) {
+  w_string_t *str = w_string_new_len(path, pathlen);
+  w_ignore_addstr(ignore, str, is_vcs_ignore);
+  w_string_delref(str);
+}
+
+void w_ignore_addstr(struct watchman_ignore *ignore, w_string_t *path,
+                     bool is_vcs_ignore) {
+  w_ht_set(is_vcs_ignore ? ignore->ignore_vcs : ignore->ignore_dirs,
+           w_ht_ptr_val(path), w_ht_ptr_val(path));
+
+  art_insert(&ignore->tree, (const unsigned char *)path->buf, path->len,
+             is_vcs_ignore ? kVCSIgnore : kFullIgnore);
+}
+
+bool w_ignore_check(struct watchman_ignore *ignore, const char *path,
+                    uint32_t pathlen) {
+  const char *skip_prefix;
+  uint32_t len;
+  art_leaf *leaf = art_longest_match(&ignore->tree, (const unsigned char *)path,
+                                     (int)pathlen);
+
+  if (!leaf) {
+    // No entry -> not ignored.
+    return false;
+  }
+
+  if (pathlen < leaf->key_len) {
+    // We wanted "buil" but matched "build"
+    return false;
+  }
+
+  if (pathlen == leaf->key_len) {
+    // Exact match.  This is an ignore if we are in kFullIgnore,
+    // but not in kVCSIgnore mode.
+    return leaf->value == kFullIgnore ? true : false;
+  }
+
+  // Our input string was longer than the leaf key string.
+  // We need to ensure that we observe a directory separator at the
+  // character after the common prefix, otherwise we may be falsely
+  // matching a sibling entry.
+  skip_prefix = path + leaf->key_len;
+  len = pathlen - leaf->key_len;
+
+  if (*skip_prefix != WATCHMAN_DIR_SEP
+#ifdef _WIN32
+      && *skip_prefix != '/'
```

Comment:
Hmm, why is this separate check needed? Is it just because people on Windows can specify ignores with both `/` and `\`? Probably worth adding a comment explaining that.